windows-gaming-driver/src/setup/vfio.rs

```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("Helpful tools to avoid this and figure out which numbers are what devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+	println!();
+	let (temp_devices, _, pci_devs) = ask_selection("Which PCI Devices do you want to pass through temporarily?", &|x| Path::new(PCI_DEVS_PATH).join(x.pci_slot.clone()).join("reset").exists(), pci_devs, false);
+	println!("Add aditional pci devices(2/2)");
+	println!("");
+	println!("If you don't know what you are doing, chose None here!");
+	println!("NOTE: EVERY PCI DEVICE LISTED HERE WILL BE PERMANENTLY BOUND TO VFIO-PCI!");
+	println!("ONLY ADD DEVICES HERE IF YOU KNOW WHAT YOU ARE DOING!");
+	println!("If you kill your system using this there is only one thing i can say to you:");
+	println!();
+	println!("U done goofed m8.");
+	println!("Ur on ur own :)");
+	println!();
+	//filter to everything except the PCI Bridges
+	let (perm_devices, perm_dev_ids, _) = ask_selection("Which PCI Device do you want to pass through permanently?", &|x| x.pci_class != "60400", pci_devs, true);
+	
+	println!("{} GPU Devices", gpu_devices.len());
+	println!("{} temp Devices", temp_devices.len());
+	println!("{} perm Devices", perm_devices.len());
+	
+	let mut all_perm_device_ids = Vec::new();
+	all_perm_device_ids.extend(gpu_dev_ids);
+	all_perm_device_ids.extend(perm_dev_ids);
+	
+	let mut all_devices = Vec::new();
+	all_devices.extend(gpu_devices);
+	all_devices.extend(temp_devices);
+	all_devices.extend(perm_devices);
+	
+	setup.vfio_devs = all_perm_device_ids;
+	machine.vfio_slots = all_devices;
+	
+    Ok(())	
+}
+
+fn ask_selection<'a>(question: &str, filter: &Fn(&PciDevice) -> (bool), pci_devices: Vec<PciDevice<'a>>, permanent: bool) -> (Vec<VfioDevice>, Vec<PciId>, Vec<PciDevice<'a>>) {
+    
+    let (mut devices, remaining_pci_devices) = pci_devices.into_two(filter);
+    
+    println!();
+    
+    for (i, dev) in devices.iter().enumerate() {
+        println!("[{}]\t{}", i, dev);
+    }
+	println!("[{}]\tNone of the above.", devices.len());
+
+    let selection = ask::numeric(question, 0..devices.len()+1);
+    
+    if selection >= devices.len() {
+	    devices.extend(remaining_pci_devices);
+    	return (Vec::new(), Vec::new(), devices);
+    }
+    let selected = devices.remove(selection);
+
+	devices.extend(remaining_pci_devices);
+	
+	let (mut related_devices, unpassed_pci_devices) = devices.into_two(&|dev| dev.pci_device() == selected.pci_device());
+	related_devices.push(selected);
+	
+	let (mut ret_devices, mut ret_devids, ret_pci_devices) = ask_selection(question, filter, unpassed_pci_devices, permanent);
+	ret_devices.extend(related_devices.iter().map(|dev| if permanent {VfioDevice::Permanent(dev.pci_slot.clone())} else {VfioDevice::Temporarily(dev.pci_slot.clone())}));
+	ret_devids.extend(related_devices.iter().map(|dev| dev.id));
+	
+	(ret_devices, ret_devids, ret_pci_devices)
+}
```

Comment:
Please restructure this.

What you're doing right now:

* Ask for GPUs (or literally anything else)
* Ask for resettable devices (or literally anything else)
* Ask for nonresettable devices (or literally anything else)

(it should be obvious why this is bad)

What you should be doing:

* Ask for a *single* GPU (or *nothing at all*)
* Ask if they would like to add other PCI devices
* Only if they answered yes, ask for any number of PCI devices (but straight up prohibit bridges because wtf) (Here, we can initially hide non-resettable devices and only show them if they specifically ask for it, but warn in that case)